opaque_keys/__init__.py

```diff
@@ -168,10 +179,18 @@ def __str__(self):
         """
         Serialize this :class:`OpaqueKey`, in the form ``<CANONICAL_NAMESPACE>:<value of _to_string>``.
         """
+        # OpaqueKeys are often repeatedly serialized, so we cache this value.
+        if self._unicode is not None:
+            return self._unicode
+
         if self.deprecated:
             # no namespace on deprecated
-            return self._to_deprecated_string()
-        return self.NAMESPACE_SEPARATOR.join([self.CANONICAL_NAMESPACE, self._to_string()])  # pylint: disable=no-member
+            self._unicode = self._to_deprecated_string()  # pylint: disable=assigning-non-slot
```

Comment:
I don't know of a particular convention (and I just saw the earlier discussion below about these comments).  Maybe just a comment somewhere in the file like "# TODO Revisit assigning-non-slot comments on pylint upgrade"?  We could then later search the code for "pylint upgrade", or stumble across it when looking for TODO comments.